Improve EH dead code elimination

BruceForstall opened the pull request:
Update `fgRemoveDeadBlocks()` to only treat EH handlers as reachable if their corresponding `try` block entry is reachable.

This causes more dead code to be eliminated in handlers that are not reachable.

There is a little more work to do to completely clean these up:
1. If a `try` is not reachable, we should get rid of its entire EH table entry. Currently, its block is generated as `int3` (on xarch).
2. Even though the handler is not reachable, we still generate the prolog followed by `int3` for the body.

Contributes to #82335

BruceForstall committed:
Improve EH dead code elimination

Update `fgRemoveDeadBlocks()` to only treat EH handlers as reachable
if their corresponding `try` block entry is reachable.

This causes more dead code to be eliminated in handlers that are
not reachable.

There is a little more work to do to completely clean these up:
1. If a `try` is not reachable, we should get rid of its entire EH table
entry. Currently, its block is generated as `int3` (on xarch).
2. Even though the handler is not reachable, we still generate the prolog
followed by `int3` for the body.

Contributes to #82335

BruceForstall commented:
@AndyAyersMS PTAL
cc @dotnet/jit-contrib

kunalspathak reviewed (approved):
LGTM

AndyAyersMS commented:
> 1. If a `try` is not reachable, we should get rid of its entire EH table entry.

We encode EH region numbers into the AddCodeDesc entries (starting in morph) so if we remove things from the EH table we'll need to fix these up too. Or perhaps rework all this so these entries hang off the EH table somehow and migrate organically along with their related regions.

BruceForstall commented:
> We encode EH region numbers into the AddCodeDsc entries

In particular, using `bbThrowIndex()`, to ensure `fgAddCodeRef` only shares throw blocks in the same EH region.

BruceForstall committed:
Fix BBJ_EHFILTERRET handling in fgRemoveBlockAsPred

fgRemoveBlockAsPred had special handling to not decrease the bbRefs count
of the filter-handler targeted by BBJ_EHFILTERRET. This is incorrect, as
the filter-handler already has an extra "beginning of handler" extra
ref count. Possibly this code was never invoked before as filters were
not previously removed.

Also, fix the JitDump output of the filter ret block in block dumps.

AndyAyersMS reviewed (approved):
Interesting that we don't see much of a TP win here.

I wonder if we should try doing this earlier?

BruceForstall commented:
> Interesting that we don't see much of a TP win here.

Would you expect a TP win because of less IR? It seems like unreachable handlers should be rare, especially after the early EH opts. This still doesn't remove an unreachable `try` or the unreachable handler (it just turns them into `int3`), or the "dead" EH region.

> I wonder if we should try doing this earlier?

Seems like it needs to be late enough for all the normal opts creating dead flow to run.

BruceForstall committed:
Put back arm32 BBJ_CALLFINALLY/BBJ_ALWAYS pair check

BruceForstall commented:
/azp run runtime-coreclr jitstress

BruceForstall commented:
Test failure is https://github.com/dotnet/runtime/issues/93506
